libraries/botbuilder-dialogs-adaptive-testing/src/testScript.ts

```diff
@@ -98,43 +111,71 @@ export class TestScript extends Configurable implements TestScriptConfiguration
+
+        const adapter = new TestAdapter(TestAdapter.createConversation(testName));
+        useBotState(adapter, userState, convoState);
+        middlewares.forEach((middleware) => {
+            adapter.use(middleware);
+        });
+
+        adapter.onTurnError = async (context, err) => {
+            await context.sendActivity(err.message);
+        };
+
+        return adapter;
+    }
+
     /**
      * Starts the execution of the test sequence.
-     * @param testName Name of the test
-     * @param testAdapter (Optional) Test adapter
+     * @param resourceExplorer The resource explorer to use.
+     * @param testName Name of the test.
+     * @param callback The bot logic.
+     * @param adapter Optional test adapter.
+     * @param middlewares Middlewares to be added to the adapter.
      */
     public async execute(
         resourceExplorer: ResourceExplorer,
         testName?: string,
-        testAdapter?: TestAdapter
+        callback?: (context: TurnContext) => Promise<void>,
+        adapter?: TestAdapter,
+        ...middlewares: Middleware[]
     ): Promise<void> {
-        if (!testAdapter) {
-            testAdapter = new TestAdapter(TestAdapter.createConversation(testName));
+        if (!adapter) {
+            adapter = this.defaultTestAdapter(testName, ...middlewares);
         }
 
-        testAdapter.enableTrace = this.enableTrace;
-        testAdapter.locale = this.locale;
+        adapter.enableTrace = this.enableTrace;
+        adapter.locale = this.locale;
 
-        const storage = new MemoryStorage();
-        const userState = new UserState(storage);
-        const convoState = new ConversationState(storage);
-        useBotState(testAdapter, userState, convoState);
+        this.userTokenMocks.forEach((userTokenMock) => {
+            userTokenMock.setup(adapter);
+        });
 
-        const bot = new DialogManager(this.dialog);
-        ResourceExtensions.useResourceExplorer(bot, resourceExplorer);
-        LanguageGeneratorExtensions.useLanguageGeneration(bot);
+        if (!callback) {
+            const dm = new DialogManager(this.dialog);
+            ResourceExtensions.useResourceExplorer(dm, resourceExplorer);
+            LanguageGeneratorExtensions.useLanguageGeneration(dm);
 
-        if (this.languagePolicy) {
-            LanguageGeneratorExtensions.useLanguagePolicy(bot, this.languagePolicy);
-        }
+            if (this.languagePolicy) {
+                LanguageGeneratorExtensions.useLanguagePolicy(dm, this.languagePolicy);
+            }
 
-        this.userTokenMocks.forEach((userTokenMock) => {
-            userTokenMock.setup(testAdapter);
-        });
+            callback = dm.onTurn.bind(dm);
+        }
 
         for (let i = 0; i < this.script.length; i++) {
             const testAction = this.script[i];
-            await testAction.execute(testAdapter, bot.onTurn.bind(bot));
+            await testAction.execute(adapter, callback);
```

Comment:
I'd rewrite this block to:

```typescript
await Promise.all(this.script.map(testAction => testAction.execute(adapter, callback)));
```

I wonder if it would also be useful to use something like [`p-map`](https://www.npmjs.com/package/p-map) and a configurable `concurrency` to control concurrent execution. What do you think?

Reply:
Is there a way to execute all the scripts in order?